src/vmm/src/arch/aarch64/regs.rs

```diff
@@ -157,265 +110,685 @@ arm64_sys_reg!(ID_AA64MMFR2_EL1, 3, 0, 0, 7, 2);
+#[derive(Default, Debug, Clone, PartialEq, Eq, Versionize)]
+struct Aarch64RegisterVecInner {
+    ids: Vec<u64>,
+    data: Vec<u8>,
 }
 
-/// Configure relevant boot registers for a given vCPU.
-///
-/// # Arguments
-///
-/// * `cpu_id` - Index of current vcpu.
-/// * `boot_ip` - Starting instruction pointer.
-/// * `mem` - Reserved DRAM for current VM.
-pub fn setup_boot_regs(
-    vcpufd: &VcpuFd,
-    cpu_id: u8,
-    boot_ip: u64,
-    mem: &GuestMemoryMmap,
-) -> Result<(), Error> {
-    let kreg_off = offset__of!(kvm_regs, regs);
-
-    // Get the register index of the PSTATE (Processor State) register.
-    let pstate = offset__of!(user_pt_regs, pstate) + kreg_off;
-    let id = arm64_core_reg_id!(KVM_REG_SIZE_U64, pstate);
-    vcpufd
-        .set_one_reg(id, PSTATE_FAULT_BITS_64.into())
-        .map_err(|err| Error::SetOneReg(id, err))?;
-
-    // Other vCPUs are powered off initially awaiting PSCI wakeup.
-    if cpu_id == 0 {
-        // Setting the PC (Processor Counter) to the current program address (kernel address).
-        let pc = offset__of!(user_pt_regs, pc) + kreg_off;
-        let id = arm64_core_reg_id!(KVM_REG_SIZE_U64, pc);
-        vcpufd
-            .set_one_reg(id, boot_ip.into())
-            .map_err(|err| Error::SetOneReg(id, err))?;
-
-        // Last mandatory thing to set -> the address pointing to the FDT (also called DTB).
-        // "The device tree blob (dtb) must be placed on an 8-byte boundary and must
-        // not exceed 2 megabytes in size." -> https://www.kernel.org/doc/Documentation/arm64/booting.txt.
-        // We are choosing to place it the end of DRAM. See `get_fdt_addr`.
-        let regs0 = offset__of!(user_pt_regs, regs) + kreg_off;
-        let id = arm64_core_reg_id!(KVM_REG_SIZE_U64, regs0);
-        vcpufd
-            .set_one_reg(id, get_fdt_addr(mem).into())
-            .map_err(|err| Error::SetOneReg(id, err))?;
-    }
-    Ok(())
+impl Aarch64RegisterVecInner {
+    /// Returns the number of elements in the vector.
+    fn len(&self) -> usize {
+        self.ids.len()
+    }
+
+    /// Returns true if the vector contains no elements.
+    fn is_empty(&self) -> bool {
+        self.ids.is_empty()
+    }
+
+    /// Appends a register to the vector, copping register data.
+    fn push(&mut self, reg: Aarch64RegisterRef<'_>) {
+        self.ids.push(reg.id);
+        self.data.extend_from_slice(reg.data);
+    }
+
+    /// Returns an iterator over stored registers.
+    fn iter(&self) -> impl Iterator<Item = Aarch64RegisterRef> {
+        Aarch64RegisterVecIterator {
+            index: 0,
+            offset: 0,
+            vec: self,
+        }
+    }
+
+    /// Returns an iterator over stored registers that allows register modifications.
+    fn iter_mut(&mut self) -> impl Iterator<Item = Aarch64RegisterRefMut> {
+        Aarch64RegisterVecIteratorMut {
+            index: 0,
+            offset: 0,
+            vec: self,
+        }
+    }
```

Comment:
nit: I think the idiomatic way of doing this in rust is to implement [`IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) for `&Aarch64RegisterVecInner` and `&mut Aarch64RegisterVecInner` respectively. That way you get for-syntax integration, too (e.g. `for item in &register_vec` instead of `for item in register_vec.iter()`)